feat(protocol): support heatpumps with new protocol E/E

tspopp opened the pull request:
# Implementing another heat pump protocol

## Help

:warning: This PR needs help by the community, as I don't have a machine with the new protocol :warning: 

## Missing Features / Open Tasks:
- [x] Checksum Calculation
- [x] MITM
  - [x] Passthrough (verified)
  - [x] Error Passthrough (verified)
  - [x] Override Values (experimental)
- [x] HMI Message
  - [x] time/date 
  - [x] target temperature
  - [x] operation types (timer/always-on)
  - [x] operation modes (boost, absence, eco...)
  - [x] air-duct config (int/int..)
  - [x] anti-legionella setting
  - [x] enable emergency mode
  - [x] disable heat-element
  - [x] enable pv-input 
  - [x] timer window a/b
  - [x] test modes
  - [x] fan exhaust config (INT/EXT)
  - [x] error request id / error number 
- [x] Main Message
  - [x] hot water temp
  - [x] input air temp
  - [x] lower evaporator temp
  - [x] upper evaporator temp
  - [x] NEW: compressor outlet temp
  - [x] states: heatpump, heatelement, ext, fan, defrost, pv-input, solar-input...
  - [x] fan pwm 
- [x] Energy Message
  - [x] total heatpump hours
  ~total hours~
  - [x] total hours heat element
  ~power heatpump~
  ~power heatelement~
  - [x] power total
  ~water production~
  - [x] total energy
  - [x] add unknown rising counters for identification
  - [x] NEW: diagnosis temperature values (min/max observed)
- [x] Error Message
  - [x] errorCode
  - [x] requestId 
  - ~optional: error message content~
- [x] Compatibility to legacy protocol
- [x] Enable Filter
- [x] MQTT Discovery depending on protocol version
- [x] Refactoring
- [x] Remove Unknown C-Value Hack

## Status Quo:

This branch contains a modified version of AquaMQTT which is meant to be installed in LISTENER mode. It is currently able to identify hmi, main and energy messages from the new protocol #45. These are provided to mqtt on the three topics:

`aquamqtt/hmi/debug`
`aquamqtt/main/debug`
`aquamqtt/energy/debug`


The new protocol findings are documented within [PROTOCOL_NEXT.md](https://github.com/tspopp/AquaMQTT/blob/t/nextProtocol/PROTOCOL_NEXT.md)

## Tracing

Using the debug python script https://github.com/tspopp/AquaMQTT/blob/main/tools/debug.py you are able to record changing messages over time and identify which location holds what kind of attribute.

### Create the python environment

```
python3 -m venv venv
source venv/bin/activate
pip install paho-mqtt
```

### Run the python script

```
source venv/bin/activate
python3 debug.py
```


## How to help here?

- start a recording of the serial messages
- change a single (previously unidentified) value at the hmi, e.g. water target temperature from 50 to 52
- stop the recording of the serial message
- attach the traces together with a description ("changed water target temperature from 50 to 52 and back")

tspopp commented:
@taloriko I've already added parsing of time and date from the hmi message. You may want to check if the values are correct.

taloriko commented:
Okay, that was my first PR that I used, but it seems to be working quite well.

These new topics were generated:

![grafik](https://github.com/user-attachments/assets/c6546552-3bfe-4279-b7dd-5e2c9109ccda)


**Check Time and Date:**

HMI set to 17 Clock --> OK
![grafik](https://github.com/user-attachments/assets/b04862c5-f973-4793-902f-560bbe874773)

HMI  set to 16 Clock --> OK
![grafik](https://github.com/user-attachments/assets/d41b97b7-12cd-46ff-ad3c-55e06b21223e)

HMI set to 30 November --> Day OK | Month --> NOK
 ![grafik](https://github.com/user-attachments/assets/87fb9cfc-3095-4a5e-b707-03e366f97602)

HMI set to 7 May --> Day OK | Month --> NOK
![grafik](https://github.com/user-attachments/assets/9efb2110-bcdd-4347-95b4-ee2fda3e61a6)

HMI set to 1 February--> Day OK | Month --> NOK
![grafik](https://github.com/user-attachments/assets/53d25c72-d869-403c-a73b-0db780e30bab)


I initially tried to locate the bytes for HMI debugging. At first glance, the byte positions are similar to the previous ones. Here is the initial list of what I was able to reproduce on the HMI in a short amount of time.

**HMI Debug**

1. Byte ????

2. Byte --> target temperature -->HEX
•  40 in Hex: 28
•  41 in Hex: 29
•  42 in Hex: 2A
•  43 in Hex: 2B
•  44 in Hex: 2C

In this dump, I set the target temperature from 40°C to 44°C in individual increments:

2228110000001000062C01D0020000000B44312310000000004E450000060422013E
2228110000001000062C01D0020000000C44312310000000004E450000060422013E
2228110000001000062C01D0020000000D44312310000000004E450000060422013E
2228110000001000062C01D0020000000E44312310000000004E450000060422013E
2228110000001000062C01D0020000000F44312310000000004E450000060422013E
2228110000001000062C01D0020000001044312310000000004E450000060422013E
2229110000001000062C01D0020000001144312310000000004E450000060422013E
2229110000001000062C01D0020000001244312310000000004E450000060422013E
2229110000001000062C01D0020000001344312310000000004E450000060422013E
222A110000001000062C01D0020000001444312310000000004E450000060422013E
222A110000001000062C01D0020000001544312310000000004E450000060422013E
222A110000001000062C01D0020000001644312310000000004E450000060422013E
222B110000001000062C01D0020000001644312310000000004E450000060422013E
222B110000001000062C01D0020000001744312310000000004E450000060422013E
222B110000001000062C01D0020000001844312310000000004E450000060422013E
222B110000001000062C01D0020000001944312310000000004E450000060422013E
222C110000001000062C01D0020000001944312310000000004E450000060422013E
222C110000001000062C01D0020000001A44312310000000004E450000060422013E
222C110000001000062C01D0020000001B44312310000000004E450000060422013E
222C110000001000062C01D0020000001C44312310000000004E450000060422013E



3. Byte --> Operating mode 

10 = Auto
11 = Eco/Manuell Eco=Aktiv
12 = Eco/Manuell Eco=Inaktiv

In this dump, I set the Operating mode from Auto --> Eco/Manuell Eco=Inaktiv --> Eco/Manuell Eco=Aktiv

2200100000001000062C01D0020000002F44313510000000004E450000060422013E
2200100000001000062C01D0020000003044313510000000004E450000060422013E
2200100000001000062C01D0020000003144313510000000004E450000060422013E
222B120000001000062C01D0020000003244313510000000004E450000060422013E
222B120000001000062C01D0020000003344313510000000004E450000060422013E
222B120000001000062C01D0020000003444313510000000004E450000060422013E
222B120000001000062C01D0020000003544313510000000004E450000060422013E
222B120000001000062C01D0020000003644313510000000004E450000060422013E
222B120000001000062C01D0020000003744313510000000004E450000060422013E
222B120000001000062C01D0020000003844313510000000004E450000060422013E
222B120000001000062C01D0020000003944313510000000004E450000060422013E
222B110000001000062C01D0020000003944313510000000004E450000060422013E
222B110000001000062C01D0020000003A44313510000000004E450000060422013E
222B110000001000062C01D0020000003B44313510000000004E450000060422013E
222B110000001000062C01D0020000000044313610000000004E450000060422013E


I hope the initial steps are okay for you.

tspopp commented:
Yes, looks great so far. I will incorporate the findings later. One more hint, if you use the python script, you will get dumps in hex and dec representation. It is WAY easier to spot the changed attributes. For example:

e.g. in dec
```
2024-11-04 18:39:09.587725,34 50 18 0 0 0 16 0 6 44 1 208 2 0 0 0 0 33 30 13 12 0 0 0 0 78 69 0 0 6 4 34 1 62
2024-11-04 18:39:10.135443,34 50 18 0 0 0 16 0 6 44 1 208 2 0 0 0 2 33 30 13 12 0 0 0 0 78 69 0 0 6 4 34 1 62
```
or in hex
```
2024-11-04 18:39:09.587725,2232120000001000062C01D00200000000211E0D0C000000004E450000060422013E
2024-11-04 18:39:10.135443,2232120000001000062C01D00200000002211E0D0C000000004E450000060422013E
```

tspopp commented:
I updated the implementation based on your findings. Please note [PROTOCOL_NEXT.md](https://github.com/tspopp/AquaMQTT/blob/t/nextProtocol/PROTOCOL_NEXT.md) . You might also try to edit that file directly or add comments directly to the document by reviewing this PR. This would be actually a nice way implementing this  :grimacing: 

 It might be helpful to also have a look on the pre-existing [PROTOCOL.md](https://github.com/tspopp/AquaMQTT/blob/main/PROTOCOL.md) since there will be a lot of similarities :)

taloriko commented:
I can't edit the [PROTOCOL_NEXT.md](https://github.com/tspopp/AquaMQTT/blob/t/nextProtocol/PROTOCOL_NEXT.md) file directly. I'm using GitHub for the first time and hope that the PR Patch 1 is suitable.

I'm currently stuck with the timer. The changes start counting from byte 0 at index 9.

I have a few examples:


```
Timer 07:00-12:00 | 16:00-22:00 | 11h

34 48 18 0 0 0 48 0 6 192 3 124 1 0 0 0 34 69 49 47 16 0 0 0 0 78 69 0 0 164 1 44 1 62

Timer 08:00-12:00 | 16:00-22:00 | 10h

34 43 17 0 0 0 16 0 6 192 3 104 1 0 0 0 7 68 49 3 18 0 0 0 0 78 69 0 0 224 1 240 0 62

Timer 09:00-13:00 | 17:00-23:00 | 10h

34 43 17 0 0 0 16 0 6 252 3 104 1 0 0 0 58 68 49 8 18 0 0 0 0 78 69 0 0 28 2 240 0 62

Timer 01:00-05:00 | 10:00-18:00 | 12h

34 48 18 0 0 0 48 0 6 88 2 224 1 0 0 0 54 69 49 27 16 0 0 0 0 78 69 0 0 60 0 240 0 62

Timer 01:00-06:00 | 10:00-18:00 | 13h

34 48 18 0 0 0 48 0 6 88 2 224 1 0 0 0 12 69 49 35 16 0 0 0 0 78 69 0 0 60 0 44 1 62

Timer 00:00-06:00 | 10:00-18:00 | 14h

34 48 18 0 0 0 48 0 6 88 2 224 1 0 0 0 38 69 49 37 16 0 0 0 0 78 69 0 0 60 0 44 1 62
```

taloriko commented:
To narrow down the results more precisely, I only changed the beginning of the first time window and observed the respective outcome.

Byte 9 could represent the minutes **since** 00:00.
Up to 4:00, the minutes fit into the first byte.
At 5:00, Byte 10 comes into play: 44 and 1 → (256*1) + 44 = 300 / 60 = 5
At 6:00, however, it no longer fits: (256*3) + 192 = 960 = 16

From 6:00, it looks like it represents minutes **until** midnight.

At 7:00, it doesn’t fit anymore – or am I calculating it incorrectly?

Starting from 6 a.m., it seems that Byte 9 and 10 switch with Byte 29 and 30, and thus the time windows change.
Or my assumption may be incorrect.

00:00-12:00 | 16:00-22:00 | 18h,22,48,18,0,0,0,48,0,2,0,0,208,2,0,0,0,40,70,49,54,16,0,0,0,0,78,69,0,0,192,3,104,1
01:00-12:00 | 16:00-22:00 | 17h,22,48,18,0,0,0,48,0,2,60,0,148,2,0,0,0,7,70,49,55,16,0,0,0,0,78,69,0,0,192,3,104,1
02:00-12:00 | 16:00-22:00 | 16h,22,48,18,0,0,0,48,0,2,120,0,88,2,0,0,0,47,70,49,55,16,0,0,0,0,78,69,0,0,192,3,104,1
03:00-12:00 | 16:00-22:00 | 15h,22,48,18,0,0,0,48,0,2,180,0,28,2,0,0,0,34,70,49,56,16,0,0,0,0,78,69,0,0,192,3,104,1
04:00-12:00 | 16:00-22:00 | 14h,22,48,18,0,0,0,48,0,2,240,0,224,1,0,0,0,15,70,49,57,16,0,0,0,0,78,69,0,0,192,3,104,1
05:00-12:00 | 16:00-22:00 | 13h,22,48,18,0,0,0,48,0,2,44,1,164,1,0,0,0,57,70,49,57,16,0,0,0,0,78,69,0,0,192,3,104,1
06:00-12:00 | 16:00-22:00 | 12h,22,48,18,0,0,0,48,0,2,192,3,104,1,0,0,0,43,70,49,58,16,0,0,0,0,78,69,0,0,104,1,104,1
07:00-12:00 | 16:00-22:00 | 11h,22,48,18,0,0,0,48,0,2,192,3,104,1,0,0,0,30,70,49,59,16,0,0,0,0,78,69,0,0,164,1,44,1
08:00-12:00 | 16:00-22:00 | 10h,22,48,18,0,0,0,48,0,2,192,3,104,1,0,0,0,24,70,49,0,17,0,0,0,0,78,69,0,0,224,1,240,0

tspopp commented:
No worries, we can focus on other attributes first. I will try to look in these dumps as soon as I have some time for this. Alternatively, if you have a formula which seems promising, we can try to implement it and see if it matches...

taloriko commented:
I’ve now tested everything I could find on the HMI (Menu + Installer Menu). I couldn’t identify any further details.

I’ve added the time windows, though I’m still not entirely certain about their full functionality.

Here are my observations:

    Bytes 3, 14, 15, 23, 24, 27, and 28 did not change during testing and were always set to 0.
    Bytes 25 and 26 consistently held the values 78 and 69 (I have no idea).
    The function of Byte 33 is still unclear (possibly a checksum).

Do you have any further ideas on how we could test these bytes? Otherwise, I would wrap up the HMI testing and shift focus to energy management.

tspopp commented:
I think initially you have identified all major attributes of interest in the hmi message :beers: A few of these leftover bytes might be reserved for commands. Checksum is the last Byte 34, which is not available in the dumps by AquaMQTT, so it is likely that there is also something else stored in Byte 33. You may try to enter the secret menu "spin the wheel left and then to the right" and you get some more advanced options. You may reverse these items as well, but AquaMQTT currently does not implement commands. I think I documented how commands work in my protocol document, not sure if they changed the pattern with the new protocol :man_shrugging: . 

But changing this advanced settings is somehow interesting, because you're changing the main message and you able to identify more information from the main message. For example, if you set the fan speed level to 55%, you will see some value changing within the main message to 55% :wink: 

In the meantime I will implement the changes to speak both protocol version at the same time, so we can merge this to main as soon as we feel like it's ready.

At some point in time I need you to provide a large (maybe a few minutes) raw dump using AquaDebug. Only the logs from AquaDebug contains the checksum and we need to find out, how the crc values are actually calculated. In MITM mode we need to recreate messages and therefore have to generate the checksum the same way. So understanding this pattern is crucial for getting MITM to work :grimacing:

tspopp commented:
No it does not need to run during the dump. There will be a lot of hmi messages and they change checksum frequently since the time is always changing :)

taloriko commented:
Today, I wanted to analyze the next set of data, and I noticed that the ESP32 restarts every 10 seconds.

I have always received the data so far, but there were occasional dropouts. I initially thought that IP-Symcon couldn’t receive the topics quickly enough, which caused these gaps.

First, I searched for a timeout issue and found one in the WiFi settings. I adjusted the value to 600:

`constexpr uint16_t WIFI_RECONNECT_CYCLE_S = 600;`

However, this change did not fix the problem.

I’m not sure if it’s related, but I also had to adjust the code slightly to be able to flash without errors:
```

#ifdef CUSTOM_CONFIGURATION
#    include "ExampleConfiguration.h"
#else
#    include "ExampleConfiguration.h"
#endif
```

I replaced CustomConfiguration with a different configuration.

tspopp commented:
Most probably I broke something with my latest commits. I will check. Sorry for that 😅

taloriko commented:
I have finally created the log files with AquaDebug.

I created various files:

    Without opening a menu
    "Super Secret Menu" opened
    Browsed errors in the "Super Secret Menu"
    "Installer Menu" opened
    
[aqua_debug_data_Normal.txt](https://github.com/user-attachments/files/17707910/aqua_debug_data_Normal.txt)
[aqua_debug_data_SuperSecret.txt](https://github.com/user-attachments/files/17707911/aqua_debug_data_SuperSecret.txt)
[aqua_debug_data_SuperSecret_Errors.txt](https://github.com/user-attachments/files/17707912/aqua_debug_data_SuperSecret_Errors.txt)
[aqua_debug_data_Installer.txt](https://github.com/user-attachments/files/17707913/aqua_debug_data_Installer.txt)

tspopp commented:
Nice, we found the error message :tada: Here is an example:

`4A410D0F2D00270015001800130000000000000000020000000000671D00000200000045110000581100007A2200000800000002000000211E000C003200000000086B`

I will add debug topics for those, but you don't need to figure out what the bytes in this message actually mean (of course you can try, if you like). AquaMQTT is currently mostly interested in forwarding these messages to the HMI controller in MITM mode and therefore we have to know how they look like.

tspopp commented:
Based on the dumps I figured out how the checksum is generated. Moreover I added error code parsing and passing. You might want to try MITM mode now and see if it works. As soon as you removed the passthrough jumper, and changed the configuation to `constexpr EOperationMode OPERATION_MODE = EOperationMode::MITM` and flashed, you may want to the test the following items to see if this is good:

- [ ] HMI controller is still getting data from the main controller (temps are shown etc)
- [ ] HMI controller is still able to controll the main controller (e.g. turn on Boost)
- [ ] HMI controller still gets a list of errors when opening the super secret menu

Be aware that overrides / controls are not yet supported for the next protocol, if the above items are working we add the functionality for sure :)

taloriko commented:
- [x] HMI controller is still getting data from the main controller (temps are shown etc)
- [x] HMI controller is still able to control the main controller (e.g. turn on Boost)
- [x] HMI controller still gets a list of errors when opening the super secret menu

The values I receive via MQTT seem correct at first glance. I can see the values on the HMI and control various components in test mode. In the super-secret menu, I can read errors.

Unfortunately, the controller restarts every 5–12 seconds.

Here are the new topics:
![{42F4341A-FB6A-4E7A-872D-922A4A58C0D2}](https://github.com/user-attachments/assets/1ce33290-e003-4e9d-a545-32047db0565b)

![{A02EED78-E47D-4FB4-9AED-A9B4655B13BD}](https://github.com/user-attachments/assets/5bb62b3c-659f-4d41-8ad3-e09391932275)

![{A4F462C9-FF74-49D0-BD3B-A4C1CE08D56C}](https://github.com/user-attachments/assets/cebbff55-4622-4fef-8df3-d042c77daf30)

tspopp commented:
Actually, these are very good news, happy to get your heat pump fully supported soon :+1: I'll do the refactoring soon and resolve the crash you have. In the meantime you should go back to listener mode, because crashes during MITM might lead in protocol dropouts. Not sure if the heat-pump likes that behavior or not :man_shrugging: You might want to proceed with completing the protocol, observing states in the main message is actually nice. These are the icons shown in the hmi controller (fan is on, heat element is on, external boiler is on...). But of course we can proceed to add things step by step after the main PR has been merged.

taloriko commented:
Okay, I’ll switch back to LISTENER mode.

As time permits, I’ll continue analyzing the main message.

tspopp commented on AquaMQTT/include/config/Configuration.h:
```diff
@@ -18,6 +18,21 @@ namespace aquamqtt
 {
+
+// TODO: remove prior to merge
+// switches solely for debugging the crash from taloriko (protocol-next) and will be removed as soon as we determined the reason for the crash
+// if this switch is enabled, no mqtt is there at all, if this already crashes, there is something wrong with the protocol parser
+// if this is not crashing, reset this back to false
+constexpr bool DEBUG_DISABLE_ENTIRE_MQTT = false;
+
+// proceed with these, disable them all (set them all to true) and start with setting a single one to false, and see if it crashes
+// if not reset them all back to true, and set the another one to false, until you find the one which is responsible for the crash (we will analyze further)
+constexpr bool DEBUG_DISABLE_MQTT_STATS_MSG = false;
+constexpr bool DEBUG_DISABLE_MQTT_HMI_MSG = false;
+constexpr bool DEBUG_DISABLE_MQTT_MAIN_MSG = false;
+constexpr bool DEBUG_DISABLE_MQTT_ENERGY_MSG = false;
+constexpr bool DEBUG_DISABLE_MQTT_ERROR_MSG = false;
```
@taloriko: Can you test the latest version? I am not sure if your crash is still present. In case it is, we have to debug a little. I added a few switches to the Configuration file. I need to locate where the issue happens on your end, since I am not able to reproduce the crash currently. Start with` DEBUG_DISABLE_ENTIRE_MQTT = true`. I expect no crashes here.

taloriko replied:
Okay, with the new version, I am encountering some errors.

![{448F1800-FBBB-4378-8310-C7D8E149A619}](https://github.com/user-attachments/assets/ad44412b-e1b7-476a-9287-6753511076ec)

tspopp replied:
This file should not be included from `Configuration.h`

```c++
#define CUSTOM_CONFIGURATION

#ifdef CUSTOM_CONFIGURATION
#include "CustomConfiguration.h"
#else
#include "ExampleConfiguration.h"
#endif
```

so you shouldn't get any build errors due to that file?

Can you check if this line is there:
`#define CUSTOM_CONFIGURATION`

tspopp replied:
Can you flash the latest and set `constexpr bool DEBUG_DISABLE_ENTIRE_MQTT = true` ... to true (within Configuration.h) ?

taloriko replied:
Yes, I have two boards: one inside the heat pump and one on the table.

Yes, nothing is connected. Switching the bridge in Listener mode also doesn’t change anything.

Previously, I had `constexpr char brokerClientId[] = "aquamqtt";` on both boards. Today, I changed this on the board on the table to aquamqttV2 and also updated the network name to `constexpr char networkName[] = "aquamqttV2";`.

Additionally, I adjusted `constexpr bool DEBUG_DISABLE_ENTIRE_MQTT = true`.

None of this changed the behavior, and it continues to reboot as before.

Since I’m still not very experienced with Visual Studio Code, here’s my process for today:

In the PlatformIO tab, I select Miscellaneous and then Clone Git Project.
There, I select` tspopp/aquamqtt`.

I create a new folder and select it as the repository destination.

When Visual Studio Code displays the folder structure, I right-click on the platformio.ini file and select Open Integrated Terminal.

In the terminal, I execute the following commands:

```
git fetch origin pull/52/head:pr-52  
git checkout pr-52
```

This is confirmed with: `Switched to branch 'pr-52'.`

Then, I copy the ExampleConfiguration.h file, rename it to Custom, and adjust my settings.

I also adjust the settings in Configuration.h.

To flash everything, I select the USB port at the bottom left in Visual Studio Code and switch to env:arduino_nano_esp32.

After that, I press the right arrow (PlatformIO upload). The upload completes successfully, and the Arduino doesn’t reboot.

However, when I press the reset button or unplug the USB cable, it starts a reboot loop.

tspopp replied:
Sorry for this, I've been finally able to reproduce this issue. I made an intermediate hack to fix the issue to get you going and test the latest state.

To run the updates you can rust run the following three commands:

```
git clone https://github.com/tspopp/AquaMQTT.git -b t/nextProtocol
pio run -e arduino_nano_esp32
pio run -t upload -e arduino_nano_esp32
```

to see the serial console you may do 

```
pio device monitor -e arduino_nano_esp32
```

I resolved the build error so you can also try with the unchanged ExampleConfiguration and see if you have no reboots there.

taloriko replied:
No need to apologize for that. Thank you for the new version!
 I just flashed your updated version, and it's running without any reboots.

**Configuration:**
```
LISTENER
constexpr bool DEBUG_DISABLE_ENTIRE_MQTT = false
```

Additionally, it created new MQTT topics for me:

![{3495EC12-F71A-48FF-B2F2-24FA710323F2}](https://github.com/user-attachments/assets/9e9d864d-b047-4489-9cb8-7a17666c8ce1)

tspopp replied:
Awesome, since we're good now and there are now further reboots, you can safely switch to MITM mode. We can proceed with implementing overrides. I'll drop you a new version soon :)

tspopp commented:
I enabled overrides for the new protocol. To test this publish some values to the MQTT control topic. For example:

On `aquamqtt/ctrl/operationMode` submit the value `BOOST` or value `ABSENCE` or an empty string to remove the override. I implemented all the overrides which are available in the legacy mode, too. Even the PV override mode shall work now. See more [here](https://github.com/tspopp/AquaMQTT/blob/main/MQTT.md#subscribe-topics).

taloriko commented:
> I enabled overrides for the new protocol (MITM only). To test this publish some values to the MQTT control topic. For example:
> 
> On `aquamqtt/ctrl/operationMode` submit the value `BOOST` or value `ABSENCE` or an empty string to remove the override. I implemented all the overrides which are available in the legacy mode, too. Even the PV override mode shall work now. See more [here](https://github.com/tspopp/AquaMQTT/blob/main/MQTT.md#subscribe-topics).
> 
> Edit: Or use the device which should appear in HomeAssistant (if you use HomeAssistant 😀 )

No, I don’t use HomeAssistant. I’m using IP-Symcon instead. Unfortunately, it’s not as convenient to set up, but we’ll manage 😆.

I briefly tested with the setup from 4 PM.

MITM is working.

Unfortunately, I don’t have much time for testing this week, but I quickly switched `$root/ctrl/operationMode` between "BOOST" and "MAN ECO ON."

The mode was correctly changed, and as a result, the "overrides" topic was set to 1.

How should I proceed with testing?
Should I check all overrides?

tspopp commented:
Yeah, if overrides are working, we are almost complete. You may just use it for a while and see if there are some serious issues before we merge this to main. Of couse you can try all of the overrides and see if it works as it should. Very nice :+1: 

What's left is identifying more energy attributes. You may want to provide a long-time trace of the energy message while your heat pump is running / heating up. The idea is, that we identify the byte locations, where the value is rising over time. Once identified, I will add something to publish them to MQTT as "unknownCounterValue1" and "unknownCounterValue2" etc. By visualizing those values you might get an idea what they are.

Silmo commented:
> Yeah, if overrides are working, we are almost complete. You may just use it for a while and see if there are some serious issues before we merge this to main. Of couse you can try all of the overrides and see if it works as it should. Very nice 👍
> 
> What's left is identifying more energy attributes. You may want to provide a long-time trace of the energy message while your heat pump is running / heating up. The idea is, that we identify the byte locations, where the value is rising over time. Once identified, I will add something to publish them to MQTT as "unknownCounterValue1" and "unknownCounterValue2" etc. By visualizing those values you might get an idea what they are.

I have an Atlantic 270/V3. Attached are the logs for a heating run from 46c to 51c.

Let me know if you need anything else.
[aquamqtt_hmi_debug_dec.csv](https://github.com/user-attachments/files/17952741/aquamqtt_hmi_debug_dec.csv)
[aquamqtt_hmi_debug_hex.csv](https://github.com/user-attachments/files/17952742/aquamqtt_hmi_debug_hex.csv)
[aquamqtt_main_debug_dec.csv](https://github.com/user-attachments/files/17952743/aquamqtt_main_debug_dec.csv)
[aquamqtt_main_debug_hex.csv](https://github.com/user-attachments/files/17952744/aquamqtt_main_debug_hex.csv)
[aquamqtt_energy_debug_dec.csv](https://github.com/user-attachments/files/17952745/aquamqtt_energy_debug_dec.csv)
[aquamqtt_energy_debug_hex.csv](https://github.com/user-attachments/files/17952746/aquamqtt_energy_debug_hex.csv)

tspopp commented:
I added a temporary hack for reviewing parsed energy values:

```
Example message from the dump
45 5 0 59 32 0 0 0 0 5 0  0  0  59  32  0  0    2  226  60  0  219  1  0  0  0  0  0  0  11  62  10  96  246  52  241  57  251  36  0  0  0  0  0  0
   a a  b  b c c c c d d  d  d   e   e  e  e    f    f   f  f    g  g                                      h   h    i   i    j   j  j  j  j  j  j  j
```

I identified a few possible values, which are parsed and put to the mqtt topics `aquamqtt/energy/unknownA` `aquamqtt/energy/unknownB` etc.

Some notes:
I guess that J might be the `totalEnergyWh`, but it is not changing during runtime. Maybe it changes every day? G is the current power drawn, but it is unclear if it will also contains the power of the heat element if both heatpump and element are on. F is some upcounting value, it could be water production but also an energy value. A, B, E, D are most likely the run time counters, but is unclear why they appear twice, as 2 byte and 4 byte value. The C value is strange, in @Silmo I saw that value downcounting, before the heat pump started :shrug: 

Looking forward how you may interprete live values from your heatpump

Silmo commented:
> I added a temporary hack for reviewing parsed energy values:
> 
> ```
> Example message from the dump
> 45 5 0 59 32 0 0 0 0 5 0  0  0  59  32  0  0    2  226  60  0  219  1  0  0  0  0  0  0  11  62  10  96  246  52  241  57  251  36  0  0  0  0  0  0
>    a a  b  b c c c c d d  d  d   e   e  e  e    f    f   f  f    g  g                                      h   h    i   i    j   j  j  j  j  j  j  j
> ```
> 
> I identified a few possible values, which are parsed and put to the mqtt topics `aquamqtt/energy/unknownA` `aquamqtt/energy/unknownB` etc.
> 
> Some notes: I guess that J might be the `totalEnergyWh`, but it is not changing during runtime. Maybe it changes every day? G is the current power drawn, but it is unclear if it will also contains the power of the heat element if both heatpump and element are on. F is some upcounting value, it could be water production but also an energy value. A, B, E, D are most likely the run time counters (run time heat element, run time heat pump), but is unclear why they appear twice, as 2 byte and 4 byte value. The C value is strange, in @Silmo dump, I saw that value downcounting, before the heat pump started 🤷
> 
> Looking forward how you may interprete live values from your heatpump
> 
> Edit: I guess another dump where the machine is running in emergency mode (heat pump disabled, heat element only) for a while might help in determining values bound to the heat element.

A & D is the run time heat element (5hrs)
[aquamqtt_energy_debug_dec.csv](https://github.com/user-attachments/files/17965875/aquamqtt_energy_debug_dec.csv)
[aquamqtt_energy_debug_hex.csv](https://github.com/user-attachments/files/17965876/aquamqtt_energy_debug_hex.csv)

B & E is the run time heat pump (8257hrs now)

Attached the log for a short run in emergency mode.
![image](https://github.com/user-attachments/assets/b64dd0e1-dcc6-4f2c-8804-23c94f5bf994)

tspopp commented:
- G is total power (sum of both heatpump and heatelement)
- C still unclear
- H, I, J have not changed. I'd hoped for change of J, since I supected it to be totalEnergyWh
- In this case I expect F to be totalEnergyWh, since that water production value we have in the legacy protocol is black magic anyway ;)

Nice to have the runtime counters identified :tada: I will change the code accordingly and leave the hack for another while.

taloriko commented:
I have a HomeMatic power measurement plug connected to the heat pump.

I can confirm that G is probably calculated Power.
![grafik](https://github.com/user-attachments/assets/4b3c62c3-e857-4648-a144-1a674486b80a)

For F, I’m more inclined to think it’s Energy in Wh.
![grafik](https://github.com/user-attachments/assets/07dd647d-9160-4a25-88a0-3f22fe94f8ec)

H, I, and J haven’t changed.
H = 10235
I = 9723
J = 6923

A and D represent the runtime of the heating element.
B and E represent the runtime of the heat pump.

I suspect that one of the counters can be reset, while the other is permanent. There is a reset option available on the HMI, which I plan to test in the coming days when I have time.
![grafik](https://github.com/user-attachments/assets/7d73d005-82fc-4974-bb86-935579d5d17f)

C started counting down from the beginning of operation.
From 2202717 to 19529728 =2492989

The final value didn’t change either after the target temperature was reached.
Only after disconnecting the heat pump from power did the value drop to 0
The peak at 6 PM is probably due to the restart.
![grafik](https://github.com/user-attachments/assets/69d28051-7854-4127-bfec-2f4b1f168908)

taloriko commented:
I share the same opinion.
Only the non-resettable counter is important as a topic.

Additionally, there are two more values in the HMI that I haven’t been able to identify yet.
I can’t tell if these are only displayed in the HMI or if they’re also transmitted:

    Operating time (Ready?)
    Number of restarts (Heat pump without power)
    
![grafik](https://github.com/user-attachments/assets/6afebd16-3619-4895-a214-7b9d9b0a78e5)

tspopp commented:
I couldn't find these attributes in the legacy protocol either. My guess is, that these values are actually calculated/counted within the HMI controller, and not transferred within the protocol :shrug: 

I've been also not able to find software version numbers in the protocol, that E/E, B/B etc.

taloriko commented:
I couldn’t resist and performed the reset of the operating hours.

A and B reset to 0 after the reset, while D and E retain their values. On the HMI, however, the values remain and don’t reset to 0.

H, I, and J can now also be linked to a display on the HMI:
![grafik](https://github.com/user-attachments/assets/d29e6d74-91c2-4e8d-b51f-78b3c913636c)

If I perform a reset here, the values change.
![grafik](https://github.com/user-attachments/assets/c2f46eeb-d0a0-4ed1-a4af-a20afba4cc4e)

taloriko commented:
Now I understand: each byte represents either a MIN or MAX temperature.

```
45 0 0 0 0 0 0 0 0 3 0 0 0 244 29 0 0 234 139 59 0 236 1 0 0 0 0 0 0 43 43 41 63 0 8 1 8 15 17 0 0 0 0 0 0
45 0 0 0 0 0 0 0 0 3 0 0 0 244 29 0 0 234 139 59 0 236 1 0 0 0 0 0 0 aa bb cc dd e f g h ii jj 0 0 0 0 0 0
```
JJ - Air Temp Max
II - Air Temp Min
h  - Evaporator Air Temp Max
g  - Evaporator Air Temp Min
f  - EvaporatorLower Air Temp Max
e  - EvaporatorLower Air Temp Min
dd - CompressoorOutlet Temp Max
cc - CompressoorOutlet Temp Max
bb - Watter Temp Max
aa - Watter Temp Min

tspopp commented:
🤯 Nice!  🎉  If you want to achieve 120% you can also dump the HMI message while triggering those resets. You will see which location is responsible for the reset 😬

taloriko commented:
I only have a phone with me right now.

On a reset, the value 2 is written:

![Screenshot_20241130_210507_Symcon.jpg](https://github.com/user-attachments/assets/6d5d534a-e05b-4cae-8731-32ceb84e729d)

tspopp commented:
I checked the statistics reset with my heatpump (legacy protocol) and unfortunately nothing changed within the serial messages. So it seems we have this values only for the heatpumps with next protocol. Are you intrested in having these values within MQTT?

taloriko commented:
I'm not sure if the values are actually needed as a topic 🤔.

I haven't used them in the past three years.
But then again, I haven't encountered any errors during that time either 😅.

Still, since we already have them, why not provide them as a topic 🤷‍♂️? There are also very few changes expected in the values.

Or perhaps something like the debug topic for self-configuration 😅.

tspopp commented:
Any new ideas regarding that C value? Else I think I will merge this within the next days :)
